File: v4/rust/iprange-livedb/src/recovery/structured_output.rs
//! Canonical output for ranges backed by recovered typed structures.

use crate::contract::{MetaV4, StructureKind};
use crate::error::Result;
use crate::immutable_output::{Builder, MembershipWords};
use crate::key::{Ipv4Key, Ipv6Key};
use crate::mapping::Mapping;
use crate::range_tree::Record;
use crate::structured_value::{NetworkEnrichmentV1, NetworkEnrichmentV1Codec};
use crate::validation::{ValidationObject, ValidationReason};

use super::direct_output::{report_overlap, DirectKey};
use super::membership_index::{Locator as MembershipLocator, MembershipIndex};
use super::range_components::Policy;
use super::report::{RecoverySink, Reporter, Unknown};
use super::structure_table::{Locator as StructureLocator, StructureIndex};
use super::tables::Tables;

#[derive(Clone, Copy)]
pub(super) struct Resolved {
    structure: StructureLocator,
    membership: Option<MembershipLocator>,
}

pub(super) struct NetworkEnrichmentV1Output<'a, 'b, S, K> {
    mapping: &'a Mapping,
    meta: MetaV4,
    memberships: &'a MembershipIndex,
    structures: &'a StructureIndex,
    tables: &'a Tables,
    builder: &'a mut Builder,
    reporter: &'a mut Reporter<'b, S>,
    marker: std::marker::PhantomData<K>,
}

impl<'a, 'b, S: RecoverySink, K: StructuredKey> NetworkEnrichmentV1Output<'a, 'b, S, K> {
    pub(super) fn new(
        mapping: &'a Mapping,
        meta: MetaV4,
        memberships: &'a MembershipIndex,
        structures: &'a StructureIndex,
        tables: &'a Tables,
        builder: &'a mut Builder,
        reporter: &'a mut Reporter<'b, S>,
    ) -> Self {
        debug_assert_eq!(structures.kind(), StructureKind::NetworkEnrichmentV1);
        Self {
            mapping,
            meta,
            memberships,
            structures,
            tables,
            builder,
            reporter,
            marker: std::marker::PhantomData,
        }
    }

    fn resolve_record(&mut self, record: Record<K>) -> Result<Option<Resolved>> {
        let Some(structure) = self.structures.get(self.tables, record.value)? else {
            self.reporter.unknown(Unknown {
                reason: ValidationReason::StructureMissing,
                object: ValidationObject::StructureDictionary,
                page_number: None,
                physical_bytes: None,
                address_fence: Some(<K as DirectKey>::fence(record.from, record.to)),
                contributes_to_possible_span: false,
                has_unbounded_extent: false,
            })?;
            return Ok(None);
        };
        let membership = if structure.membership_id == 0 {
            None
        } else {
            let Some(membership) = self.memberships.get(self.tables, structure.membership_id)?
            else {
                self.reporter.unknown(Unknown {
                    reason: ValidationReason::StructureMembershipInvalid,
                    object: ValidationObject::StructureDictionary,
                    page_number: Some(structure.leaf_page),
                    physical_bytes: None,
                    address_fence: Some(<K as DirectKey>::fence(record.from, record.to)),
                    contributes_to_possible_span: false,
                    has_unbounded_extent: false,
                })?;
                return Ok(None);
            };
            Some(membership)
        };
        Ok(Some(Resolved {
            structure,
            membership,
        }))
    }

    fn push(&mut self, record: Record<K>, resolved: Resolved) -> Result<()> {
        let value = NetworkEnrichmentV1Codec::decode(&resolved.structure.payload)?;
        match resolved.membership {
            Some(membership) => {
                let words = membership.words(self.mapping, self.meta);
                <K as StructuredKey>::push(
                    self.builder,
                    record.from,
                    record.to,
                    value,
                    Some(&words),
                )
            }
            None => <K as StructuredKey>::push::<super::membership_words::RecoveredWords<'_>>(
                self.builder,
                record.from,
                record.to,
                value,
                None,
            ),
        }
    }
}

impl<S: RecoverySink, K: StructuredKey> Policy<K> for NetworkEnrichmentV1Output<'_, '_, S, K> {
    type Resolved = Resolved;

    fn resolve(&mut self, record: Record<K>) -> Result<Option<Self::Resolved>> {
        self.resolve_record(record)
    }

    fn accept(&mut self, record: Record<K>, resolved: Option<Self::Resolved>) -> Result<()> {
        let Some(resolved) = resolved else {
            return self.reporter.ranges_rejected(1, record.from, record.to);
        };
        self.push(record, resolved)?;
        self.reporter.range_accepted(record.from, record.to)
    }

    fn reject_overlap(&mut self, count: u64, from: K, to: K) -> Result<()> {
        report_overlap(self.reporter, count, from, to)
    }

    fn finish(&mut self) -> Result<()> {
        Ok(())
    }
}

pub(crate) trait StructuredKey: DirectKey {
    fn push<W: MembershipWords>(
        builder: &mut Builder,
        from: Self,
        to: Self,
        value: NetworkEnrichmentV1,
        membership: Option<&W>,
    ) -> Result<()>;
}

impl StructuredKey for Ipv4Key {
    fn push<W: MembershipWords>(
        builder: &mut Builder,
        from: Self,
        to: Self,
        value: NetworkEnrichmentV1,
        membership: Option<&W>,
    ) -> Result<()> {
        builder.push_network_enrichment_v1_v4(from, to, value, membership)
    }
}

impl StructuredKey for Ipv6Key {
    fn push<W: MembershipWords>(
        builder: &mut Builder,
        from: Self,
        to: Self,
        value: NetworkEnrichmentV1,
        membership: Option<&W>,
    ) -> Result<()> {
        builder.push_network_enrichment_v1_v6(from, to, value, membership)
    }
}

File: v4/go/internal/recovery/structured_output.go
package recovery

// Canonical output for ranges backed by recovered typed structures
// (Rust recovery/structured_output.rs): one accepted component pushes
// the decoded network-enrichment value with its optional membership
// bitmap, and one unresolved structure streams the missing-structure
// or invalid-membership fence.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/validation"
	"github.com/firehol/iprange/v4/go/internal/writer"
)

// structuredResolved is one resolved structure record with its
// optional membership (Rust Resolved).
type structuredResolved struct {
	structure  structureLocator
	membership *membershipLocator
}

// structuredOutput is the structured-range policy of the
// overlap-component pass (Rust NetworkEnrichmentV1Output).
type structuredOutput struct {
	mapping     *mapping.Mapping
	meta        format.Meta
	memberships *membershipIndex
	structures  *structureIndex
	tables      *tableStore
	builder     *writer.OutputBuilder
	rep         *reporter
	family      uint8
}

// resolve proves one record's structure and membership (Rust
// resolve_record: the structure lookup, the membership dependency, and
// their fence envelopes).
func (o *structuredOutput) resolve(record rangeRecord) (any, error) {
	structure, found, err := o.structures.get(o.tables, record.value)
	if err != nil {
		return nil, err
	}
	if !found || structure.rejected {
		fence := o.codec().fence(record.from, record.to)
		if err := o.rep.unknown(unknownEnvelope{
			reason:       validation.ReasonStructureMissing,
			object:       validation.ObjectStructureDictionary,
			addressFence: &fence,
		}); err != nil {
			return nil, err
		}
		return nil, nil
	}
	var membership *membershipLocator
	if structure.membershipID != 0 {
		locator, found, err := o.memberships.get(o.tables, structure.membershipID)
		if err != nil {
			return nil, err
		}
		if !found || locator.rejected {
			fence := o.codec().fence(record.from, record.to)
			page := structure.leafPage
			if err := o.rep.unknown(unknownEnvelope{
				reason:       validation.ReasonStructureMembershipInvalid,
				object:       validation.ObjectStructureDictionary,
				pageNumber:   &page,
				addressFence: &fence,
			}); err != nil {
				return nil, err
			}
			return nil, nil
		}
		membership = &locator
	}
	return structuredResolved{structure: structure, membership: membership}, nil
}

// accept counts one accepted component and pushes it (Rust
// NetworkEnrichmentV1Output::accept: the missing structure reports the
// rejected range, the resolved structure streams the range).
func (o *structuredOutput) accept(record rangeRecord, resolved any) error {
	value, ok := resolved.(structuredResolved)
	if !ok {
		if err := o.codec().reportRejected(o.rep, 1, record.from, record.to); err != nil {
			return err
		}
		return nil
	}
	payload := value.structure.payloadBytes()
	decoded, err := format.DecodeNetworkEnrichmentV1(payload)
	if err != nil {
		return err
	}
	if err := o.push(record, decoded, value.membership); err != nil {
		return err
	}
	return o.codec().reportAccepted(o.rep, record)
}

// rejectOverlap streams one whole overlap component (Rust
// NetworkEnrichmentV1Output::reject_overlap).
func (o *structuredOutput) rejectOverlap(count uint64, from, to rangeKey) error {
	return reportOverlap(o.rep, o.codec(), count, from, to)
}

// finish proves nothing for the structured policy (Rust
// NetworkEnrichmentV1Output::finish: no pending range).
func (o *structuredOutput) finish() error { return nil }

// push streams one structured range to the destination builder (Rust
// StructuredKey::push over the optional membership words).
func (o *structuredOutput) push(record rangeRecord, value format.NetworkEnrichmentV1, membership *membershipLocator) error {
	switch o.family {
	case format.AddressFamilyIPv4:
		if membership == nil {
			return o.builder.PushNetworkEnrichmentV1V4Words(uint32(record.from.hi), uint32(record.to.hi), value, nil)
		}
		return o.builder.PushNetworkEnrichmentV1V4Words(uint32(record.from.hi), uint32(record.to.hi), value, locatorWords{reader: membershipWordReader{m: o.mapping, meta: o.meta, locator: *membership}})
	case format.AddressFamilyIPv6:
		if membership == nil {
			return o.builder.PushNetworkEnrichmentV1V6Words(record.from.hi, record.from.lo, record.to.hi, record.to.lo, value, nil)
		}
		return o.builder.PushNetworkEnrichmentV1V6Words(record.from.hi, record.from.lo, record.to.hi, record.to.lo, value, locatorWords{reader: membershipWordReader{m: o.mapping, meta: o.meta, locator: *membership}})
	default:
		return corruptError("recovery structured output family is invalid")
	}
}

// codec returns the family codec of one structured output.
func (o *structuredOutput) codec() rangeCodec {
	switch o.family {
	case format.AddressFamilyIPv4:
		return rangeV4Codec{}
	case format.AddressFamilyIPv6:
		return rangeV6Codec{}
	default:
		return nil
	}
}
